oeps/oep-0012.rst

```diff
@@ -0,0 +1,305 @@
+
+* The "Discussion" tab (aka the discussion board) for course-wide discussions
+
+* A discussion that is shown within the "Teams" tab for each course team
+
+* In the future, there are thoughts of having program-wide discussions,
+  enterprise-wide discussions, and global discussions with a social group
+
+* There is also the concept of changing course announcements to be a new type of
+  discussion post
+
+In an ideal world, Open edx Discussions would be a separately-installable
+feature that would provide plugins for all of the functionality listed above. An
+Open edX operator could choose not to install edX discussions and choose to
+install different collaboration features instead.
+
+Currently the discussion board is a Django view built into edx-platform, but it
+should instead be made into a plugin. One option would be to create a new
+CourseDiscussionXBlock to display the "Course" tab, but as described above there
+are downsides to forcing this to be an XBlock. It would be cleaner if the edX
+discussion library provided a variety of UI blocks for each of the ways in which
+discussion content can be rendered. These blocks would share functionality
+between them, but would also be customized for their unique context. The data
+for the discussions would be managed through a REST API.
+
+An interesting question is how can these UI blocks share course-scoped settings
+that the instructor specified in Studio.
+`OEP-6 Context-scoped XBlock Fields`_.
+
+Use Case 3: Instructor Dashboard
+--------------------------------
+
+The Instructor Dashboard provides a set of user interfaces to allow an
+instructor to manage the features of their course. Currently these user
+interfaces must be added directly to edx-platform, even if the feature
+itself lives in a Python library that has been separately installed.
+
+It would be preferable if any feature implemented as a Python library could
+contribute new sections to the Instructor Dashboard through plugins. The
+dashboard would show a list of the feature that can be managed, and then
+cliking on a feature would present the user with the pluggable interface.
+These interfaces could be made available for instructors in the LMS as
+today, or a new instructor portal could be implemented reusing the same
+blocks.
+
+Specification
+=============
+
+Requirements
+------------
+
+What are the requirements for a pluggable user interface?
+
+1. **It should render its content into a section of a page**
+
+   A pluggable user interface should be able to return the HTML, CSS and custom
+   JavaScript necessary to render its content. In addition, it should be able to
+   declare the JavaScript libraries that it depends upon.
+
+   Python and JavaScript APIs will be provided so that the consumer of a UI
+   block can render it. These APIs will ensure that the necessary libraries have
+   been loaded, and that the necessary CSS and JavaScript files have been added
+   to the page (without duplication).
```

Comment:
The "without duplication" part here will be interesting to implement. UIBlocks will have to both be self-contained and namespace all of their own dependencies so they don't conflict with each other, but also communicate somehow to resolve that only one instance of any one library version needs to be loaded.

Maybe this is an implementation detail, but I'm already thinking about how to handle this - do we make the page serve a common bundle, and then each UI block requests only their unique dependencies as a separate bundle? Or are all UIBlock dependencies on a page hoisted into one megabundle?